### ISSUE TITLE
Angle predicted only as 0 or 90

### ISSUE THREAD

**Issue by midnightdim** [ultralytics/ultralytics#16657]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I've trained a model using yolo11s-obb.pt on a dataset of 800 PCB images (50 in validation set) containing capacitors at various orientations (with images taken from different angles). While the object detection works well, the angle predictions are consistently around 0 or 90 degrees, regardless of the actual orientation in the images.

What could be the cause of this? My dataset includes rotated elements at various angles, and I've also tried rotation augmentation without success.

Attached are images showing initial labeling and predictions:

![val_batch2_labels](https://github.com/user-attachments/assets/728e49c7-ea19-40d3-a145-4cc73f52e052)
![val_batch2_pred](https://github.com/user-attachments/assets/bf9f5c31-f767-4674-94a0-281e158d81e8)

As you can see, the labels reflect the correct angles, but the predictions do not.

Any suggestions are welcome. Thank you!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @midnightdim, thank you for reaching out to Ultralytics 🚀! We appreciate your detailed report and images. An Ultralytics engineer will assist you soon.

In the meantime, here are a few steps you can take which might shed light on the issue:

- Ensure your dataset and annotations are correctly formatted and include a diverse range of angles.
- Consider verifying your training setup and hyperparameters align with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
- If this is a 🐛 Bug Report, a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) would be valuable for us to investigate further.

Don't forget to check out the [Docs](https://docs.ultralytics.com) for any further insights, which include many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.

Feel free to join our community discussions. You can get real-time support in our [Discord](https://ultralytics.com/discord) 🎧, engage in detailed conversations on [Discourse](https://community.ultralytics.com), or participate in our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Make sure you've upgraded to the latest `ultralytics` package:

```bash
pip install -U ultralytics
```

This ensures you benefit from the latest fixes and features.

## Environments

YOLO runs in various environments. Try these verified ones to see if the issue persists:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** and **Amazon** deep learning setups or explore our [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge reflects the status of Ultralytics CI testing on multiple platforms, ensuring proper functionality.

Stay connected with the community and share your insights. We're here to help! 😊

**Comment by Y-T-G** [comment]
Can you post your training command?

**Comment by midnightdim** [comment]
> Can you post your training command?

Sure, it's trivial:
```
model = YOLO('yolo11s-obb.pt')
model.train(data='datasets/data.yaml', epochs=50)
```

**Comment by Y-T-G** [comment]
What size are your images?

**Comment by midnightdim** [comment]
> What size are your images?

1024 x 768 px

**Comment by Y-T-G** [comment]
Try training at a different `imgsz`. Like 512

**Comment by Y-T-G** [comment]
Related https://github.com/ultralytics/ultralytics/issues/15573

**Comment by midnightdim** [comment]
> Try training at a different `imgsz`. Like 512

I tried 512. Almost the same result. But I noticed that for one item in my tests it detected it at 97 degrees. Previously all values were within the range of 0-2 or 88-92 degrees.

**Comment by Y-T-G** [comment]
What's you mAP after training?

**Comment by Y-T-G** [comment]
You can try a different loss function.

1. Install `mmcv`:
```
pip install openmim
mim install mmcv
```
2, Train using this code:
```python
from mmcv.ops import diff_iou_rotated_2d

def rotated_iou_loss(pred, target, linear=False, mode='log', eps=1e-6):
    """Rotated IoU loss.

    Computing the IoU loss between a set of predicted rbboxes and target
     rbboxes.
    The loss is calculated as negative log of IoU.

    Args:
        pred (torch.Tensor): Predicted bboxes of format (x, y, h, w, angle),
            shape (n, 5).
        target (torch.Tensor): Corresponding gt bboxes, shape (n, 5).
        linear (bool, optional): If True, use linear scale of loss instead of
            log scale. Default: False.
        mode (str): Loss scaling mode, including "linear", "square", and "log".
            Default: 'log'
        eps (float): Eps to avoid log(0).
    Return:
        torch.Tensor: Loss tensor.
    """
    assert mode in ['linear', 'square', 'log']
    if linear:
        mode = 'linear'
        warnings.warn(
            'DeprecationWarning: Setting "linear=True" in '
            'poly_iou_loss is deprecated, please use "mode=`linear`" '
            'instead.')

    if diff_iou_rotated_2d is None:
        raise ImportError('Please install mmcv-full >= 1.5.0.')

    ious = diff_iou_rotated_2d(pred.unsqueeze(0), target.unsqueeze(0))
    ious = ious.squeeze(0).clamp(min=eps)

    if mode == 'linear':
        loss = 1 - ious
    elif mode == 'square':
        loss = 1 - ious**2
    elif mode == 'log':
        loss = -ious.log()
    else:
        raise NotImplementedError
    return loss

from ultralytics import YOLO
import ultralytics.utils.loss as loss

loss.probiou = rotated_iou_loss

model = YOLO('yolo11s-obb.pt')

model.train(data='datasets/data.yaml', epochs=50)
```

**Comment by Y-T-G** [comment]
You can try a different loss function.

1. Install `mmcv`:
```
pip install openmim
mim install mmcv
```
2, Train using this code:
```python
from mmcv.ops import diff_iou_rotated_2d

def rotated_iou_loss(pred, target, linear=False, mode='vanilla', eps=1e-6):
    """Rotated IoU loss.

    Computing the IoU loss between a set of predicted rbboxes and target
     rbboxes.
    The loss is calculated as negative log of IoU.

    Args:
        pred (torch.Tensor): Predicted bboxes of format (x, y, h, w, angle),
            shape (n, 5).
        target (torch.Tensor): Corresponding gt bboxes, shape (n, 5).
        linear (bool, optional): If True, use linear scale of loss instead of
            log scale. Default: False.
        mode (str): Loss scaling mode, including "linear", "square", and "log".
            Default: 'log'
        eps (float): Eps to avoid log(0).
    Return:
        torch.Tensor: Loss tensor.
    """
    assert mode in ['linear', 'square', 'log', 'vanilla']
    if linear:
        mode = 'linear'
        warnings.warn(
            'DeprecationWarning: Setting "linear=True" in '
            'poly_iou_loss is deprecated, please use "mode=`linear`" '
            'instead.')

    if diff_iou_rotated_2d is None:
        raise ImportError('Please install mmcv-full >= 1.5.0.')

    pred[:, [2, 3]] = pred[:, [3, 2]]
    target[:, [2, 3]] = target[:, [3, 2]]
    ious = diff_iou_rotated_2d(pred.unsqueeze(0), target.unsqueeze(0)).nan_to_num_(nan=0.0)
    ious = ious.squeeze(0).clamp(min=eps)

    if mode == 'linear':
        loss = 1 - ious
    elif mode == 'square':
        loss = 1 - ious**2
    elif mode == 'log':
        loss = -ious.log()
    elif mode == "vanilla":
        loss = ious
    else:
        raise NotImplementedError
    return loss

from ultralytics import YOLO
import ultralytics.utils.loss as loss

loss.probiou = rotated_iou_loss

model = YOLO('yolo11s-obb.pt')

model.train(data='datasets/data.yaml', epochs=50, box=0.01)
```

**Comment by midnightdim** [comment]
> What's you mAP after training?

mAP50 (all) is 0.993

Can it be related to the fact that capacitors and mounting holes are round on these images?

**Comment by glenn-jocher** [comment]
A high mAP50 suggests strong detection performance, but the round shapes of capacitors and mounting holes could indeed affect angle prediction due to ambiguity in orientation. You might consider augmenting your dataset with more varied rotations or ensuring the model receives sufficient training examples at diverse angles.
